src/Cache/ShardedPhpRedis.php

```diff
@@ -0,0 +1,326 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\redis\Cache\PhpRedis.
+ */
+
+namespace Drupal\redis\Cache;
+
+use Drupal\Core\Cache\Cache;
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Cache\CacheTagsChecksumInterface;
+
+/**
+ * PhpRedis cache backend.
+ */
+class ShardedPhpRedis extends CacheBase {
```

Comment:
Not sure. I am using multi/pipeline for multiple cache get/set. But I'm not sure if that would make a difference that's big enough to be worth maintaining two separate implementations.

I did a quick experiment to just copy your implementation over mine, here's what I noticed:
* My last flush/deleteAll implementation keeps the value in a object property, to avoid repeated fetchs for it... I think that's a worthy optimization, otherwise we have to do an additional get for ever cache get?
* As mentioned, I'm using a pipeline to do multiple cache gets
* I have a separate method expandEntry() to convert the redis hash array into a cache object and check if it's still valid. I think that helps to keep get() shorter and easier to understand.
* Similar for set(), I have a createEntryHash(). I'm totally fine with renaming them, but I think they are useful.
* In get(), you immediately delete invalid entries. Is that really worth it? I think we can make the assumption that someone will actually rebuild and set that cache almost immediately anyway?
* A Cache::validateTags($tags); call in set()
* A small optimization to do multiple delete by passing an array of keys to get(), not sure if that is shardable? 
* You usually have the real implementation in the single method, in my case it's in multiple, I don't really care which way round it is.
* I implemented invalidateAll() using a cache-tag-per-bin, that then works with the allow invalid, but I'd be OK with dropping that.

So not that much different and a few of those might be useful to merge into your implementation?

Note: About the allow invalid: IMHO, that is mostly useful for tag based invalidation, there I can imagine real use cases (like having render cache support for a min-age, so that a rendered piece of content is used for at least 10 minutes). For things that are explicitly cached with an expiration date, I don't care that much...